Improve detection for invalid billing and shipping addresses in the Checkout block

ralucaStan opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

Closes https://linear.app/a8c/issue/WOOPLUG-5264/block-checkout-the-shipping-address-section-remains-collapsed-even
Closes https://github.com/woocommerce/woocommerce/issues/60222
Closes https://linear.app/a8c/issue/WOOPLUG-5264/block-checkout-the-shipping-address-section-remains-collapsed-even
Closes https://github.com/woocommerce/woocommerce/issues/58622

(For Bug Fixes) Bug introduced in PR # .




<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

## Admin user
In your testing, you'll need to use an admin account, that has saved **complete & different** billing and shipping addresses; Place an oder on the store to make sure this happens. You'll need to be always logged in.

1. Add a product to cart and go to the Checkout block
2. Make sure both addresses are collapsed (not showing forms)
3. Go to your admin profile (Click top right the user, Edit profile) and remove the zip code for the billing address
4. Go back to checkout, refresh the page and see that the form is expanded, with the zip code missing
5. Add a zip code an place the order
6. Go to the user profile in the admin and notice the billing address has a zip code

Repeat the same tests with the shipping address.

## Guest user
1. Add a product to cart and go to the Checkout block
2. Make sure both addresses are expanded (showing forms)
3. Fill parts of the shipping address
4. Refresh page and notice it's not collapsed
6. Add all fields and refresh
7. Notice the form collapsed
10. Do the 3 - 7 steps with the billing address
11. Click edit for billing address and remove a field, notice the inline error
12. Refresh, notice it's collapsed (the full saved address is used, expected behaviour)
13. Place order
14. Check that the data is what you entered on the Order confirmation screen

## Programmatic testing - taken from https://github.com/woocommerce/woocommerce/pull/51386
1. Open Checkout on guest, forms should be empty, fill out your shipping address.
2. From console, run `wp.data.dispatch('wc/store/checkout').setEditingShippingAddress(false);`
3. Shipping address should collapse.
4. Run `wp.data.dispatch('wc/store/checkout').setEditingShippingAddress(true);`.
5. Shipping address should expand.
6. Fill out the billing address.
7. Run `wp.data.dispatch('wc/store/checkout').setEditingBillingAddress(false);`
9. It should collapse.
10. Run `wp.data.dispatch('wc/store/checkout').setEditingBillingAddress(true);`
11. Address should expand.
12. Remove fields from billing steps, so that an error is shown to you.
15. Run `wp.data.dispatch('wc/store/checkout').setEditingBillingAddress(false);`
16. Nothing should happen.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

ralucaStan committed:
Get real time validation errors for billing and shipping addresses when they change

ralucaStan committed:
Add changelog

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan committed:
Fix lint errors

ralucaStan committed:
Add testing

ralucaStan committed:
Merge branch 'wooplug-5264-block-checkout-the-shipping-address-section-remains' of github.com:woocommerce/woocommerce into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan committed:
Update logic for is expanded state

ralucaStan committed:
Update implementation to detect user data being loaded and add tests

ralucaStan committed:
Merge branch 'wooplug-5264-block-checkout-the-shipping-address-section-remains' of github.com:woocommerce/woocommerce into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/base/context/hooks/use-validation.ts:
```diff
@@ -34,24 +34,19 @@ export const useValidation = (): ValidationData => {
 			const store = select( validationStore );
-			return {
-				hasValidationErrors: store.hasValidationErrors(),
-				getValidationError: ( validationErrorId: string ) =>
```
A new function reference for `getValidationError` was created on every render, causing warnings.

opr replied:
This will regress. Please take a look at https://github.com/woocommerce/woocommerce/pull/60385 which explains why the callback approach won't work

ralucaStan replied:
removed in https://github.com/woocommerce/woocommerce/pull/60441/commits/348581fa8208ad157070ce72382137f70fb2fc4d

ralucaStan commented on plugins/woocommerce/client/blocks/assets/js/blocks/checkout/inner-blocks/checkout-shipping-address-block/test/customer-address.tsx:
```diff
@@ -0,0 +1,421 @@
+			'false'
+		);
+	} );
+} );
```
the email is not part of the billing form, that's why is was removed from the check; it has a separate field outside of the sections

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains

opr commented:
For guest users step 2

> Make sure both addresses are collapsed (not showing forms)

should this be the other way round, forms should be expanded since there's no address?

ralucaStan commented:
> should this be the other way round, forms should be expanded since there's no address?

nice catch, I've updated the step

opr commented on plugins/woocommerce/client/blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/test/customer-address.tsx:
```diff
@@ -0,0 +1,440 @@
+	) )
+);
+
+jest.mock( '../../../address-card', () =>
+	jest.fn( ( { isExpanded } ) => (
+		<div data-testid="address-card">
+			<div data-testid="is-expanded">{ isExpanded.toString() }</div>
+		</div>
+	) )
+);
```
I feel like this is technically OK since these components are no the systems under test, but what if we _don't_ mock these components and instead test the real ones are rendered, it could potentially catch other bugs that this method won't?

ralucaStan replied:
I think, for the purpose of testing this functionality, we can keep the mocks. The components themselves should have tests to catch other bugs.

opr commented on plugins/woocommerce/client/blocks/assets/js/blocks/checkout/inner-blocks/checkout-billing-address-block/customer-address.tsx:
```diff
@@ -21,39 +25,43 @@ const CustomerAddress = () => {
+				( key ) => key !== 'email'
+			);
+			// Check if any billing field has validation errors
+			return billingFieldKeys.some( ( key ) => {
+				const error = store.getValidationError( 'billing_' + key );
+				return error !== undefined;
+			} );
```
I'm not sure if this should be _inside_ the `useSelect`  was the selector returned by the previous implementation returning stale validation data? Is it time we create a selector to get all validation errors so that we can iterate over them more easily? It would probably solve the issue in #60385 too

ralucaStan replied:
There is no need for that part to be inside the `useSelect`, it's actually better outside to prevent any side effects.


> it time we create a selector to get all validation errors so that we can iterate over them more easily? It would probably solve the issue in https://github.com/woocommerce/woocommerce/pull/60385 too

According [to the docs](https://developer.wordpress.org/news/2024/03/how-to-work-effectively-with-the-useselect-hook/#but-call-them-outside-when-you-re-in-an-event-handler), we can use something like this: 

```
const { getValidationError } = useSelect( validationStore );
```
and we'd get the callback without any reactivity; I've tested it and it works, no warnings in the console; 

But I do like the idea of having a separate selector for the errors object. This way, we're parsing the error keys instead of the whole billing address keys. So I'll go with that.

ralucaStan replied:
Here is the logic using the selector above: 

```

const { getValidationError } = useSelect( validationStore );

const billingFieldKeys = Object.keys( billingAddress ).filter(
( key ) => key !== 'email'
);

useEffect( () => {
		// Check if any billing field has validation errors
		const hasValidationErrors = billingFieldKeys.some( ( key ) => {
			const error = getValidationError( 'billing_' + key );
			return error !== undefined;
		} );

		// Forces editing state if store has errors,
		// but not on initial render when all fields are empty.
		if (
			isInitialized &&
			hasValidationErrors &&
			editingBillingAddress === false
		) {
			setEditingBillingAddress( true );
		}
	}, [ editingBillingAddress, billingAddress, isInitialized ] );
```

What I don't like about this approach is that using `useSelect` like this is not common knowledge, and it can lead to confusion, or people subscribing to other keys without reactivity.

opr replied:
> What I don't like about this approach is that using useSelect like this is not common knowledge, and it can lead to confusion, or people subscribing to other keys without reactivity.

An inline comment could help here. I also wasn't aware of this but I like it! I'd prefer us to go with this approach.

ralucaStan replied:
I've read your comment after I made changes in https://github.com/woocommerce/woocommerce/pull/60441/commits/a1592f1f3ecb5fc224afa52ff100e1dbfb7e9cdf to use a new selector. 

Let me know what you think, I like this approach more.

opr reviewed (changes requested):
Thanks for working on this @ralucaStan it seems to be working well, I just left some inline comments that I'd like you to address before merging. The `useValidation` one definitely needs to be changed. I'd say let's not introduce any big changes to the validation data store in this PR, but maybe address them separately and refactor this one when they're merged.

BTW, I don't think the data store changes should block this, but we should consider doing them.

ralucaStan committed:
Revert useValidation change

ralucaStan committed:
Add a new selector to get all validation errors

ralucaStan committed:
Improve detection of validation errors using the new selector

opr commented on plugins/woocommerce/client/blocks/assets/js/data/validation/selectors.ts:
```diff
@@ -36,3 +36,11 @@ export const getValidationErrorId = ( state: State, errorId: string ) => {
+ * Gets all validation errors.
+ *
+ * @param { State } state The current state.
+ * @return { Record<string, any> } All validation errors.
```
This is not correct, it would be `Record< string, FieldValidationStatus >`

ralucaStan replied:
Updated. And also added an entry in the documentation

opr commented on plugins/woocommerce/client/blocks/assets/js/blocks/checkout/inner-blocks/checkout-shipping-address-block/customer-address.tsx:
```diff
@@ -21,40 +25,53 @@ const CustomerAddress = () => {
+		const hasValidationErrors = Object.keys( validationErrors ).some(
+			( key ) => {
+				if (
+					key.startsWith( 'shipping_' ) &&
```
Do we want to cross-reference this with the list of fields in `ADDRESS_FORM_KEYS` ? if someone adds an error called `shipping_selection` for example, it would trigger this too

opr replied:
You could potentially do this by iterating over `ADDRESS_FORM_KEYS` and checking if there's a validation error with `shipping_${ key }` as the name.

ralucaStan replied:
I checked for the shippingAddress/billingAddress keys, because those have the additional fields as well.

opr reviewed (changes requested):
The new selector is good, I think it's going to be useful, should we add docs for it? (Maybe as a follow up is fine too)

I left a comment about some minor improvements

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains

ralucaStan committed:
Update selector

ralucaStan committed:
Check for shippingAddress/billingAddress keys and find them in the validation object.
The shippingAddress/billingAddress also has additional fields, and does not include the email field.

ralucaStan committed:
Fix linting

opr commented on plugins/woocommerce/client/blocks/assets/js/data/validation/selectors.ts:
```diff
@@ -36,3 +41,11 @@ export const getValidationErrorId = ( state: State, errorId: string ) => {
 export const hasValidationErrors = ( state: State ) => {
 	return Object.keys( state ).length > 0;
 };
+
```
Without the import, we get a complaint:

`ESLint: The type 'FieldValidationStatus' is undefined. (jsdoc/no-undefined-types)`

With the import we also get a complaint 🙈 `TS6133: FieldValidationStatus is declared but its value is never read.`

I think we could actually change this to just use TS annotations it would be better.

```ts
/**
 * Gets all validation errors.
 */
export const getValidationErrors = (
	state: State
): Record< string, FieldValidationStatus > => state;
```

I don't think it's important to describe the argument as these selectors should not be called outside of a `@wordpress/data` context anyway.

opr reviewed (approved):
This is working well for me, but I did notice one minor issue with TS/JSDoc, so you can address that and merge. Thanks for working on it!

ralucaStan committed:
Remove comment for selector

ralucaStan committed:
Merge branch 'trunk' into wooplug-5264-block-checkout-the-shipping-address-section-remains
